### ISSUE TITLE
`DB::Exception: The specified key does not exist. This error happened for S3 disk`

### ISSUE THREAD

**Issue by kssenii** [ClickHouse/ClickHouse#65914]
`02413_replace_partition_zero_copy`:
https://s3.amazonaws.com/clickhouse-test-reports/65625/6236d9d66fa3c1d1207bfc2849d7a65f9eef728f/stateless_tests__asan__[4_4].html

cc @alesapin

**Comment by kssenii** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/65625/8c861e9c950e456030a7d73b40f72b05f725c761/stateless_tests__tsan__s3_storage__[1_4].html

**Comment by kssenii** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/65625/ec5542db137a9ca97884bd54d73a17410cba0134/stateless_tests__release_.html

**Comment by tavplubix** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/0/e749b0852a0fb672e84e3e20b126fdc4038512bb/stateless_tests__tsan__s3_storage__[2_3].html

**Comment by CheSema** [comment]
I did a research about this logs [https://s3.amazonaws.com/clickhouse-test-reports/0/e749b0852a0fb672e84e3e20b126fdc4038512bb/stateless_tests__tsan__s3_storage__[2_3].html](https://s3.amazonaws.com/clickhouse-test-reports/0/e749b0852a0fb672e84e3e20b126fdc4038512bb/stateless_tests__tsan__s3_storage__%5B2_3%5D.html)

I see here a race condition between:
`StorageReplicatedMergeTree::executeLogEntry` -> `StorageReplicatedMergeTree::attachPartHelperFoundValidPart` -> `IMergeTreeDataPart::loadColumnsChecksumsIndexes` -> `IMergeTreeDataPart::loadColumns`

https://github.com/ClickHouse/ClickHouse/blob/89de7cc1e697be43535c167e37bcb977f291502c/src/Storages/MergeTree/IMergeTreeDataPart.cpp#L1656

and

`StorageReplicatedMergeTree::executeDropRange` -> `IMergeTreeDataPart::makeCloneInDetached` -> `DataPartStorageOnDiskBase::freeze`

https://github.com/ClickHouse/ClickHouse/blob/89de7cc1e697be43535c167e37bcb977f291502c/src/Storages/MergeTree/DataPartStorageOnDiskBase.cpp#L521




When Clickhouse apply `DROP RANGE` from replication log it does not safe operations:
- move part's directory to the detached
- and removing some data in already moved parts.
The other thread could be trying reading that part right between this two operations. The read happens in not safe way too:
- firstly is checked that `if (existsFile(name))`
- and only after is performed `return readFile(name, settings, read_hint, file_size);`
If the file is removed from metadata and object storage than it could pass the check `if (existsFile(name))` and got error when file is read.

**Comment by CheSema** [comment]
<details>
  <summary>Also there are a lot of less severe errors like this</summary>
```
2025.01.22 07:21:14.163005 [ 1141 ] {} <Error> test_1sd4jojc.alter_table1 (51884811-610e-4d59-8f48-5cafbb6b1b03): Part /var/lib/clickhouse/disks/s3/store/518/51884811-610e-4d59-8f48-5cafbb6b1b03/detached/all_107_107_0/ is broken and needs manual correction. Reason: Code: 226. DB::Exception: No count.txt in part all_118_118_0. (NO_FILE_IN_DATA_PART), Stack trace
 (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:106: Poco::Exception::Exception(String const&, int) @ 0x0000000022186220
1. ./build_docker/./src/Common/Exception.cpp:106: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000010775f78
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007e8381e
3. DB::Exception::Exception<String const&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x0000000008b95c00
4. ./build_docker/./src/Storages/MergeTree/IMergeTreeDataPart.cpp:1372: DB::IMergeTreeDataPart::loadRowsCount() @ 0x000000001c88be78
5. ./build_docker/./src/Storages/MergeTree/IMergeTreeDataPart.cpp:821: DB::IMergeTreeDataPart::loadColumnsChecksumsIndexes(bool, bool) @ 0x000000001c8881d3
6. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:2236: DB::StorageReplicatedMergeTree::attachPartHelperFoundValidPart(DB::ReplicatedMergeTreeLogEntry const&) const @ 0x000000001c320277
7. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:2308: DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x000000001c321488
8. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:3875: bool std::__function::__policy_invoker<bool (std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<DB::StorageReplicatedMergeTree::processQueueEntry(std::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>)::$_1, bool (std::sh
ared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&) @ 0x000000001c441381
9. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001cdcd9ca
10. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:3871: DB::StorageReplicatedMergeTree::processQueueEntry(std::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x000000001c37537b
11. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:3934: bool std::__function::__policy_invoker<bool ()>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<DB::StorageReplicatedMergeTree::scheduleDataProcessingJob(DB::BackgroundJobsAssignee&)::$_0, bool ()>>(std::__function::__policy_storage const*) @ 0x000000001c4423a9
12. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001c47780e
13. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:308: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::routine(std::shared_ptr<DB::TaskRuntimeData>) @ 0x000000001c97675f
14. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:353: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::threadFunction() @ 0x000000001c977490
15. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:57: void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::MergeTreeBackgroundExecutor(String, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics
::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::basic_string_view<char, std::char_traits<char>>)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000001c97ce22
16. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001088be32
17. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:312: ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::T
hreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'()::operator()() @ 0x0000000010893886
18. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:344: void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGl
obalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x00000000108937a2
19. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001088896e
20. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:312: void* std::__thread_proxy[abi:ne180100]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void (ThreadPoolImpl<std::thread>::ThreadFromThreadPool::*)(), ThreadPoolImpl<std::thread>::ThreadFromThreadPool*>>(void*) @ 0x0000000010890afc
21. __tsan_thread_start_func @ 0x0000000007df6428
22. ? @ 0x00007f8571f9bac3
23. ? @ 0x00007f857202d850
 (version 25.1.1.3968 (official build))
```

```
2025.01.22 07:21:21.216870 [ 1136 ] {} <Error> test_1sd4jojc.alter_table1 (51884811-610e-4d59-8f48-5cafbb6b1b03): Part /var/lib/clickhouse/disks/s3/store/518/51884811-610e-4d59-8f48-5cafbb6b1b03/detached/all_150_150_0/ is broken and needs manual correction. Reason: Code: 226. DB::Exception: No columns.txt in part all_166_166_0, expected path store/518/51884811-610e-4d59-8f48-5cafbb6b1b03/detached/all_150_150_0/columns.txt on drive s3. (NO_FILE_IN_DATA_PART), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:106: Poco::Exception::Exception(String const&, int) @ 0x0000000022186220
1. ./build_docker/./src/Common/Exception.cpp:106: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000010775f78
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007e8381e
3. DB::Exception::Exception<String const&, String&, String>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<String&>::type, std::type_identity<String>::type>, String const&, String&, String&&) @ 0x00000000176880ec
4. ./build_docker/./src/Storages/MergeTree/IMergeTreeDataPart.cpp:1611: DB::IMergeTreeDataPart::loadColumns(bool) @ 0x000000001c88a3fd
5. ./build_docker/./src/Storages/MergeTree/IMergeTreeDataPart.cpp:811: DB::IMergeTreeDataPart::loadColumnsChecksumsIndexes(bool, bool) @ 0x000000001c887f33
6. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:2236: DB::StorageReplicatedMergeTree::attachPartHelperFoundValidPart(DB::ReplicatedMergeTreeLogEntry const&) const @ 0x000000001c320277
7. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:2308: DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x000000001c321488
8. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:3875: bool std::__function::__policy_invoker<bool (std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<DB::StorageReplicatedMergeTree::processQueueEntry(std::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>)::$_1, bool (std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&) @ 0x000000001c441381
9. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001cdcd9ca
10. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:3871: DB::StorageReplicatedMergeTree::processQueueEntry(std::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x000000001c37537b
11. ./build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:3934: bool std::__function::__policy_invoker<bool ()>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<DB::StorageReplicatedMergeTree::scheduleDataProcessingJob(DB::BackgroundJobsAssignee&)::$_0, bool ()>>(std::__function::__policy_storage const*) @ 0x000000001c4423a9
12. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001c47780e
13. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:308: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::routine(std::shared_ptr<DB::TaskRuntimeData>) @ 0x000000001c97675f
14. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:353: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::threadFunction() @ 0x000000001c977490
15. ./build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:57: void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::MergeTreeBackgroundExecutor(String, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::basic_string_view<char, std::char_traits<char>>)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000001c97ce22
16. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001088be32
17. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:312: ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'()::operator()() @ 0x0000000010893886
18. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:344: void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne180100]<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x00000000108937a2
19. ./contrib/llvm-project/libcxx/include/__functional/function.h:714: ? @ 0x000000001088896e
20. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:312: void* std::__thread_proxy[abi:ne180100]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void (ThreadPoolImpl<std::thread>::ThreadFromThreadPool::*)(), ThreadPoolImpl<std::thread>::ThreadFromThreadPool*>>(void*) @ 0x0000000010890afc
21. __tsan_thread_start_func @ 0x0000000007df6428
22. ? @ 0x00007f8571f9bac3
23. ? @ 0x00007f857202d850
 (version 25.1.1.3968 (official build))
```
</details>


This happens when Clickhouse scans all detached parts and tries to find the one which has the same checksums. Meanwhile this part is moved to the active set by other thread. The file with metadata is moved to other place.

**Comment by CheSema** [comment]
I do not understand how the error could happen in this logs
https://s3.amazonaws.com/clickhouse-test-reports/65625/ec5542db137a9ca97884bd54d73a17410cba0134/stateless_tests__release_.html

It all happened inside one query ([e40ba54e-4955-4fa2-b0c0-7179faaea295](https://pastila.nl/?00252664/d0f8be2e2a5918ba051cbb0bc171ea24#xcoaTi2ULWQh0r0LQX5ZDg==))
The file `metadata_version.txt` is removed and written again

```
158ae6d79ffc    2024-09-18      2024-09-18 09:37:32     2024-09-18 09:37:32.843288      Upload  e40ba54e-4955-4fa2-b0c0-7179faaea295    9757    VFSWrite        s3_disk test    s3/saq/tdsugvrtqputxtnpnrndamhbumdiu            1       
158ae6d79ffc    2024-09-18      2024-09-18 09:37:32     2024-09-18 09:37:32.850158      Delete  e40ba54e-4955-4fa2-b0c0-7179faaea295    68841   TCPHandler      s3_disk test    s3/saq/tdsugvrtqputxtnpnrndamhbumdiu    store/373/3735926f-f153-4a35-9e7d-4f6f105302a2/tmp_replace_from_all_1_1_0/metadata_version.txt  1       
158ae6d79ffc    2024-09-18      2024-09-18 09:37:32     2024-09-18 09:37:32.854902      Upload  e40ba54e-4955-4fa2-b0c0-7179faaea295    9766    VFSWrite        s3_disk test    s3/jly/dduuopogroobxmakqfrgduyvrkpzf    store/373/3735926f-f153-4a35-9e7d-4f6f105302a2/tmp_replace_from_all_1_1_0/metadata_version.txt  1       
```

But somehow when the data part is initialized from disk the old s3-key has been read at the method `loadColumnsChecksumsIndexes`.

I do not know why or how it is possible.

**Comment by CheSema** [comment]
https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/json.html?PR=31083&sha=c448aa135bc711bd4bd2fcd3675e76a4af21c778&name_0=PR&name_1=Stateless%20tests%20%28asan%2C%20s3%20storage%2C%202%2F3%29

```
2025.05.26 16:47:24.067105 [ 1120 ] {} <Trace> deleteFileFromS3: Writing Delete operation for blob test/tjx/cldskacpckznqccogbzyzwabpdbls
2025.05.26 16:47:24.067186 [ 1120 ] {} <Debug> deleteFileFromS3: Object with path test/tjx/cldskacpckznqccogbzyzwabpdbls was removed from S3
2025.05.26 16:47:24.165851 [ 1120 ] {} <Error> void DB::MetadataOperationsHolder::commitImpl(const TransactionCommitOptionsVariant &, SharedMutex &): Code: 107. DB::ErrnoException: Cannot open file /var/lib/clickhouse/disks/s3/store/c5c/c5c2a4e6-38d1-4efe-93f1-c53ca6c9c322/detached/all_59_59_0/data.packed: , errno: 2, strerror: No such file or directory. (FILE_DOESNT_EXIST), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x000000003b233a51
1. ./ci/tmp/build/./src/Common/Exception.cpp:110: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000001d9ee8b1
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000c959c4b
3. ./src/Common/Exception.h:242: DB::ErrnoException::ErrnoException<String&>(int, int, FormatStringHelperImpl<std::type_identity<String&>::type>, String&) @ 0x000000001dbe4572
4. ./src/Common/Exception.h:257: void DB::ErrnoException::throwFromPath<String&>(int, String const&, FormatStringHelperImpl<std::type_identity<String&>::type>, String&) @ 0x000000001dbe3a98
5. ./ci/tmp/build/./src/IO/WriteBufferFromFile.cpp:51: DB::WriteBufferFromFile::WriteBufferFromFile(String const&, unsigned long, int, std::shared_ptr<DB::Throttler>, unsigned int, char*, unsigned long, bool, unsigned long) @ 0x000000001debd03e
6. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:634: std::__unique_if<DB::WriteBufferFromFile>::__unique_single std::make_unique[abi:ne190107]<DB::WriteBufferFromFile, std::filesystem::path, unsigned long&, int&, std::shared_ptr<DB::Throttler> const&, int, std::nullptr_t, int, bool const&, unsigned long const&>(std::filesystem::path&&, unsigned long&, int&, std::shared_ptr<DB::Throttler> const&, int&&, std::nullptr_t&&, int&&, bool const&, unsigned long const&) @ 0x000000002a736153
7. ./ci/tmp/build/./src/Disks/DiskLocal.cpp:350: DB::DiskLocal::writeFile(String const&, unsigned long, DB::WriteMode, DB::WriteSettings const&) @ 0x000000002a7228c1
8. ./ci/tmp/build/./src/Disks/ObjectStorages/MetadataStorageFromDiskTransactionOperations.cpp:280: DB::WriteFileOperation::execute(std::unique_lock<DB::SharedMutex>&) @ 0x000000002a8bc6d9
9. ./ci/tmp/build/./src/Disks/ObjectStorages/MetadataOperationsHolder.cpp:65: DB::MetadataOperationsHolder::commitImpl(std::variant<std::monostate, DB::MetaInKeeperCommitOptions<zkutil::ZooKeeper>, DB::MetaInKeeperCommitOptions<DB::ZooKeeperWithFaultInjection>> const&, DB::SharedMutex&) @ 0x000000002a8c43ec
10. ./ci/tmp/build/./src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:853: void std::__function::__policy_invoker<void (unsigned long)>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<DB::DiskObjectStorageTransaction::writeFile(String const&, unsigned long, DB::WriteMode, DB::WriteSettings const&, bool)::$_0, void (unsigned long)>>(std::__function::__policy_storage const*, unsigned long) @ 0x000000002a8118cb
11. ./ci/tmp/build/./src/IO/WriteBuffer.cpp:95: DB::WriteBuffer::finalize() @ 0x000000001dc4f7ca
12. ./ci/tmp/build/./src/IO/PackedFilesWriter.cpp:217: DB::PackedFilesWriter::finalize(std::vector<String, std::allocator<String>> const&) @ 0x000000001df518d0
13. ./ci/tmp/build/./src/Storages/MergeTree/DataPartStorageOnDiskPacked.cpp:529: DB::DataPartStorageOnDiskPacked::finalizeWriter() @ 0x0000000031b1f7b3
14. ./ci/tmp/build/./src/Storages/MergeTree/DataPartStorageOnDiskPacked.cpp:397: DB::DataPartStorageOnDiskPacked::commitTransaction() @ 0x0000000031b200a5
15. ./ci/tmp/build/./src/Storages/MergeTree/IMergeTreeDataPart.cpp:1300: DB::IMergeTreeDataPart::writeMetadataVersion(std::shared_ptr<DB::Context const>, int, bool) @ 0x0000000031775008
16. ./ci/tmp/build/./src/Storages/StorageReplicatedMergeTree.cpp:2277: DB::StorageReplicatedMergeTree::attachPartHelperFoundValidPart(DB::ReplicatedMergeTreeLogEntry const&) const @ 0x0000000030c6c776
17. ./ci/tmp/build/./src/Storages/StorageReplicatedMergeTree.cpp:2340: DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x0000000030c6e194
18. ./ci/tmp/build/./src/Storages/StorageReplicatedMergeTree.cpp:3909: bool std::__function::__policy_invoker<bool (std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<DB::StorageReplicatedMergeTree::processQueueEntry(std::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>)::$_1, bool (std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&) @ 0x0000000030e12cfa
19. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x00000000320e6239
20. ./ci/tmp/build/./src/Storages/StorageReplicatedMergeTree.cpp:3905: DB::StorageReplicatedMergeTree::processQueueEntry(std::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x0000000030ce628f
21. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x0000000031745d14
22. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:318: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::routine(std::shared_ptr<DB::TaskRuntimeData>) @ 0x000000003191d288
23. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:363: DB::MergeTreeBackgroundExecutor<DB::RoundRobinRuntimeQueue>::threadFunction() @ 0x000000003191ec7d
24. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x000000001dce2893
25. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:117: ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'()::operator()() @ 0x000000001dcf09b0
26. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x000000001dcdcf0e
27. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:117: void* std::__thread_proxy[abi:ne190107]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void (ThreadPoolImpl<std::thread>::ThreadFromThreadPool::*)(), ThreadPoolImpl<std::thread>::ThreadFromThreadPool*>>(void*) @ 0x000000001dceb6b0
28. asan_thread_start(void*) @ 0x000000000c90ce77
29. ? @ 0x0000000000094ac3
30. ? @ 0x0000000000126850
 (version 25.6.2.1)
2025.05.26 16:47:24.167103 [ 1120 ] {} <Error> test_2x398bm9.alter_table1 (c5c2a4e6-38d1-4efe-93f1-c53ca6c9c322): auto DB::StorageReplicatedMergeTree::processQueueEntry(ReplicatedMergeTreeQueue::SelectedEntryPtr)::(anonymous class)::operator()(LogEntryPtr &) const: Code: 107. DB::ErrnoException: Cannot open file /var/lib/clickhouse/disks/s3/store/c5c/c5c2a4e6-38d1-4efe-93f1-c53ca6c9c322/detached/all_59_59_0/data.packed: , errno: 2, strerror: No such file or directory: While committing metadata operation #0. (FILE_DOESNT_EXIST), Stack trace (when copying this message, always include the lines below):
<the same stack trace>
```

`9b8ee31c-ee55-4942-b651-e760311e8a6f`
`02443_detach_attach_partition`
```
2025.05.26 16:47:24.063891 [ 16084 ] {9b8ee31c-ee55-4942-b651-e760311e8a6f} <Debug> executeQuery: (from [::1]:45066) (comment: 02443_detach_attach_partition.sh) (query 1, line 1) ALTER TABLE alter_table1 ATTACH PARTITION ID 'all'; (stage: Complete)
....
2025.05.26 16:47:24.360885 [ 16084 ] {9b8ee31c-ee55-4942-b651-e760311e8a6f} <Debug> test_2x398bm9.alter_table1 (c5c2a4e6-38d1-4efe-93f1-c53ca6c9c322): Checking part attaching_all_59_59_0
2025.05.26 16:47:24.363264 [ 16084 ] {9b8ee31c-ee55-4942-b651-e760311e8a6f} <Debug> AWSClient: Response status: 404, Not Found
2025.05.26 16:47:24.364411 [ 16084 ] {9b8ee31c-ee55-4942-b651-e760311e8a6f} <Debug> ReadBufferFromS3: Caught exception while reading S3 object. Bucket: test, Key: test/tjx/cldskacpckznqccogbzyzwabpdbls, Version: Latest, Offset: 0, Attempt: 1/4, Message: Code: 499. DB::Exception: The specified key does not exist. This error happened for S3 disk. (S3_ERROR) (version 25.6.2.1)
2025.05.26 16:47:24.471960 [ 16084 ] {9b8ee31c-ee55-4942-b651-e760311e8a6f} <Error> executeQuery: Code: 499. DB::Exception: The specified key does not exist. This error happened for S3 disk.: while reading key: test/tjx/cldskacpckznqccogbzyzwabpdbls, from bucket: test: Cache info: Buffer path: test/tjx/cldskacpckznqccogbzyzwabpdbls, hash key: 1aab38b8d62044e74bc2983213ba2fa3, file_offset_of_buffer_end: 0, read_until_position: 914, internal buffer end: 0, read_type: REMOTE_FS_READ_AND_PUT_IN_CACHE, last caller: 9b8ee31c-ee55-4942-b651-e760311e8a6f:16084, file segment info: File segment: [0, 913], key: 1aab38b8d62044e74bc2983213ba2fa3, state: DOWNLOADING, downloaded size: 0, reserved size: 0, downloader id: 9b8ee31c-ee55-4942-b651-e760311e8a6f:16084, current write offset: 0, caller id: 9b8ee31c-ee55-4942-b651-e760311e8a6f:16084, kind: Regular, unbound: 0. (S3_ERROR) (version 25.6.2.1) (from [::1]:45066) (comment: 02443_detach_attach_partition.sh) (query 1, line 1) (in query: ALTER TABLE alter_table1 ATTACH PARTITION ID 'all';), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x000000003b233a51
1. ./ci/tmp/build/./src/Common/Exception.cpp:110: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000001d9ee8b1
2. ./src/Common/Exception.h:118: DB::S3Exception::S3Exception(String const&, Aws::S3::S3Errors) @ 0x0000000028f2b5ee
3. ./ci/tmp/build/./src/IO/ReadBufferFromS3.cpp:449: DB::ReadBufferFromS3::sendRequest(unsigned long, unsigned long, std::optional<unsigned long>) const @ 0x0000000028f35f3a
4. ./ci/tmp/build/./src/IO/ReadBufferFromS3.cpp:404: DB::ReadBufferFromS3::initialize(unsigned long) @ 0x0000000028f325c5
5. ./ci/tmp/build/./src/IO/ReadBufferFromS3.cpp:130: DB::ReadBufferFromS3::nextImpl() @ 0x0000000028f31676
6. DB::ReadBuffer::next() @ 0x000000000cb93acb
7. ./ci/tmp/build/./src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:973: DB::CachedOnDiskReadBufferFromFile::nextImplStep() @ 0x0000000028c06ddb
8. ./ci/tmp/build/./src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:811: DB::CachedOnDiskReadBufferFromFile::nextImpl() @ 0x0000000028c0552f
9. DB::ReadBuffer::next() @ 0x000000000cb93acb
10. ./ci/tmp/build/./src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:120: DB::ReadBufferFromRemoteFSGather::readImpl() @ 0x000000002a86bb6f
11. ./ci/tmp/build/./src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:92: DB::ReadBufferFromRemoteFSGather::nextImpl() @ 0x000000002a86ba0b
12. DB::ReadBuffer::next() @ 0x000000000cb93acb
13. ./ci/tmp/build/./src/Disks/IO/ThreadPoolRemoteFSReader.cpp:147: DB::ThreadPoolRemoteFSReader::execute(DB::IAsynchronousReader::Request, bool) @ 0x00000000288f8a5b
14. ./ci/tmp/build/./src/Disks/IO/ThreadPoolRemoteFSReader.cpp:115: DB::ThreadPoolRemoteFSReader::execute(DB::IAsynchronousReader::Request) @ 0x00000000288f950f
15. ./ci/tmp/build/./src/Disks/IO/AsynchronousBoundedReadBuffer.cpp:121: DB::AsynchronousBoundedReadBuffer::readSync(char*, unsigned long) @ 0x0000000028c1e184
16. ./ci/tmp/build/./src/Disks/IO/AsynchronousBoundedReadBuffer.cpp:240: DB::AsynchronousBoundedReadBuffer::nextImpl() @ 0x0000000028c204e4
17. DB::ReadBuffer::next() @ 0x000000000cb93acb
18. ./src/IO/ReadBuffer.h:96: DB::ReadBuffer::readStrict(char*, unsigned long) @ 0x000000001dbf1e6b
19. ./src/IO/ReadHelpers.h:118: DB::PackedFilesReader::readIndex(DB::ReadBuffer&) @ 0x000000001df44de6
20. ./ci/tmp/build/./src/IO/PackedFilesReader.cpp:22: DB::PackedFilesReader::PackedFilesReader(std::shared_ptr<DB::IDisk>, String const&, DB::ReadSettings const&) @ 0x000000001df44416
21. ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:41: DB::DataPartStorageOnDiskPacked::resetReader(DB::ReadSettings const&) @ 0x0000000031b0cd9c
22. ./ci/tmp/build/./src/Storages/MergeTree/DataPartStorageOnDiskPacked.cpp:49: DB::DataPartStorageOnDiskPacked::DataPartStorageOnDiskPacked(std::shared_ptr<DB::IVolume>, String, String, DB::ReadSettings const&, bool) @ 0x0000000031b0d5fc
23. ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:41: DB::DataPartStorageOnDiskPacked* std::construct_at[abi:ne190107]<DB::DataPartStorageOnDiskPacked, std::shared_ptr<DB::IVolume> const&, String const&, String const&, DB::ReadSettings const&, bool&, DB::DataPartStorageOnDiskPacked*>(DB::DataPartStorageOnDiskPacked*, std::shared_ptr<DB::IVolume> const&, String const&, String const&, DB::ReadSettings const&, bool&) @ 0x0000000031b0b379
24. ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:49: DB::MergeTreeDataPartBuilder::getPartStorageByType(DB::MergeTreeDataPartStorageType, std::shared_ptr<DB::IVolume> const&, String const&, String const&, bool, DB::ReadSettings const&) @ 0x0000000031b07b77
25. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataPartBuilder.cpp:163: DB::MergeTreeDataPartBuilder::getPartStorageAndMarkType(std::shared_ptr<DB::IVolume> const&, String const&, String const&, DB::ReadSettings const&) @ 0x0000000031b09037
26. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataPartBuilder.cpp:182: DB::MergeTreeDataPartBuilder::withPartFormatFromVolume() @ 0x0000000031b09b95
27. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataPartBuilder.cpp:176: DB::MergeTreeDataPartBuilder::withPartFormatFromDisk() @ 0x0000000031b0971a
28. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeData.cpp:7419: DB::MergeTreeData::tryLoadPartsToAttach(std::shared_ptr<DB::IAST> const&, bool, std::shared_ptr<DB::Context const>, DB::MergeTreeData::PartsTemporaryRename&) @ 0x0000000031a2440a
29. ./ci/tmp/build/./src/Storages/StorageReplicatedMergeTree.cpp:7063: DB::StorageReplicatedMergeTree::attachPartition(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, bool, std::shared_ptr<DB::Context const>) @ 0x0000000030d4d4f3
30. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeData.cpp:6170: DB::MergeTreeData::alterPartition(std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::vector<DB::PartitionCommand, std::allocator<DB::PartitionCommand>> const&, std::shared_ptr<DB::Context const>) @ 0x0000000031a01120
31. ./ci/tmp/build/./src/Interpreters/InterpreterAlterQuery.cpp:304: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x000000002c7f388b
```

looks the same

**Comment by CheSema** [comment]
I see a problem in `StorageReplicatedMergeTree::executeLogEntry` for `LogEntry::ATTACH_PART`
https://github.com/ClickHouse/ClickHouse/blob/90b1f878b8c3f41c2a21ed7298b042c30d2a7c8f/src/Storages/StorageReplicatedMergeTree.cpp#L2264

`attachPartHelperFoundValidPart` iterates through all parts in detached directory.
It loads all the parts. Without any locking.
If part does not have `checksums.txt`, it writes it. Right into the `detached/part_name`. Without locking.
If `part_checksum` is different than part is skipped.
if `part_checksum` is equal than `metadata_version.txt` is rewritten right in `detached/part_name`. Without locking.

Metadata files could disappear while part is being loaded. Like this:
```
2025.01.22 07:21:14.163005 [ 1141 ] {} <Error> test_1sd4jojc.alter_table1 (51884811-610e-4d59-8f48-5cafbb6b1b03): Part /var/lib/clickhouse/disks/s3/store/518/51884811-610e-4d59-8f48-5cafbb6b1b03/detached/all_107_107_0/ is broken and needs manual correction. Reason: Code: 226. DB::Exception: No count.txt in part all_118_118_0. (NO_FILE_IN_DATA_PART), Stack trace 
```

Resources are spend on calculating `checksums.txt`,  may be fetch it faster? With zero-copy it is faster.
Adding file `checksums.txt` in `detached/part_name` could have a race. One update of this file could be forgotten and leaved on S3. 
For packed parts it rewrites `data.packed` file. Rewriting is read and write. Read could end up with 404 in case of race condition.

Rewriting file `metadata_version.txt` in `detached/part_name` also have a race condition. 
404 we see at the logs here.
Other operation `alter table attach partition all` could not read the part, because `metadata_version.txt` data has been rewritten/deleted. For packed parts it rewrites `data.packed` file. All the rest data is affected as well.

**Comment by CheSema** [comment]
Even after https://github.com/ClickHouse/ClickHouse/pull/81232
could be 404 errors on S3 presented in logs.

At `StorageReplicatedMergeTree::executeLogEntry` for each detached part a file `checksums.txt` is read. That part could be moved and the file is rewritten in different thread in different command, for example `alter table attach partition all.
That type of race condition here is acceptable, such part is just ignored. But nonetheless a error message is written to the logs.

There are other cases, which triggers 404 at the logs: when object storage checks that remote path does not exists. 404 response is expected [here](https://github.com/ClickHouse/clickhouse-private/issues/23229).

We need some way to mute such logging for a specific scope.

**Comment by CheSema** [comment]
Also I suspect that rewrite operation could lose data when transaction is rolled back.
https://github.com/ClickHouse/ClickHouse/pull/81066#discussion_r2123791699

Fix: https://github.com/ClickHouse/ClickHouse/pull/81787

**Comment by CheSema** [comment]
I close this.
If we have S3_ERROR No such key in CI lets open new issue with more narrow diagnostics.
This issue covered several critical cases already and it becomes incomprehensible to understand them all.
